**_La clave del éxito en cualquier organización reside en atraer y retener al mejor talento.
Este análisis es útil para un analista de RR. HH., ya que su tarea consiste en determinar qué factores retienen a los empleados en la empresa y cuáles motivan la salida de otros.
Al conocer estos factores, el analista de RR. HH. puede adoptar medidas para evitar la pérdida de personal cualificado._**

<div style="text-align: center; background-color: #856ff8; padding: 10px;">
    <h2 style="font-weight: bold;">DESCRIPCIÓN</h2>
</div>

| Name                       | Description                                                                                                                                                    |   |   |   |
|----------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------|---|---|---|
| AGE                        | Numerical Value                                                                                                                                                |   |   |   |
| ATTRITION                  | Employee leaving the company (0=no, 1=yes)                                                                                                                     |   |   |   |
| BUSINESS TRAVEL            | (1=No Travel, 2=Travel Frequently, 3=Tavel Rarely)                                                                                                             |   |   |   |
| DAILY RATE                 | Numerical Value - Salary Level                                                                                                                                 |   |   |   |
| DEPARTMENT                 | (1=HR, 2=R&D, 3=Sales)                                                                                                                                         |   |   |   |
| DISTANCE FROM HOME         | Numerical Value - THE DISTANCE FROM WORK TO HOME                                                                                                               |   |   |   |
| EDUCATION                  | Numerical Value                                                                                                                                                |   |   |   |
| EDUCATION FIELD            | (1=HR, 2=LIFE SCIENCES, 3=MARKETING, 4=MEDICAL SCIENCES, 5=OTHERS, 6= TEHCNICAL)                                                                               |   |   |   |
| EMPLOYEE COUNT             | Numerical Value                                                                                                                                                |   |   |   |
| EMPLOYEE NUMBER            | Numerical Value - EMPLOYEE ID                                                                                                                                  |   |   |   |
| ENVIROMENT SATISFACTION    | Numerical Value - SATISFACTION WITH THE ENVIROMENT                                                                                                             |   |   |   |
| GENDER                     | (1=FEMALE, 2=MALE)                                                                                                                                             |   |   |   |
| HOURLY RATE                | Numerical Value - HOURLY SALARY                                                                                                                                |   |   |   |
| JOB INVOLVEMENT            | Numerical Value - JOB INVOLVEMENT                                                                                                                              |   |   |   |
| JOB LEVEL                  | Numerical Value - LEVEL OF JOB                                                                                                                                 |   |   |   |
| JOB ROLE                   | (1=HC REP, 2=HR, 3=LAB TECHNICIAN, 4=MANAGER, 5= MANAGING DIRECTOR, 6= REASEARCH DIRECTOR, 7= RESEARCH SCIENTIST, 8=SALES EXECUTIEVE, 9= SALES REPRESENTATIVE) |   |   |   |
| JOB SATISFACTION           | Numerical Value - SATISFACTION WITH THE JOB                                                                                                                    |   |   |   |
| MARITAL STATUS             | (1=DIVORCED, 2=MARRIED, 3=SINGLE)                                                                                                                              |   |   |   |
| MONTHLY INCOME             | Numerical Value - MONTHLY SALARY                                                                                                                               |   |   |   |
| MONTHY RATE                | Numerical Value - MONTHY RATE                                                                                                                                  |   |   |   |
| NUMCOMPANIES WORKED        | Numerical Value - NO. OF COMPANIES WORKED AT                                                                                                                   |   |   |   |
| OVER 18                    | (1=YES, 2=NO)                                                                                                                                                  |   |   |   |
| OVERTIME                   | (1=NO, 2=YES)                                                                                                                                                  |   |   |   |
| PERCENT SALARY HIKE        | Numerical Value - PERCENTAGE INCREASE IN SALARY                                                                                                                |   |   |   |
| PERFORMANCE RATING         | Numerical Value - ERFORMANCE RATING                                                                                                                            |   |   |   |
| RELATIONS SATISFACTION     | Numerical Value - RELATIONS SATISFACTION                                                                                                                       |   |   |   |
| STANDARD HOURS             | Numerical Value - STANDARD HOURS                                                                                                                               |   |   |   |
| STOCK OPTIONS LEVEL        | Numerical Value - STOCK OPTIONS                                                                                                                                |   |   |   |
| TOTAL WORKING YEARS        | Numerical Value - TOTAL YEARS WORKED                                                                                                                           |   |   |   |
| TRAINING TIMES LAST YEAR   | Numerical Value - HOURS SPENT TRAINING                                                                                                                         |   |   |   |
| WORK LIFE BALANCE          | Numerical Value - TIME SPENT BEWTWEEN WORK AND OUTSIDE                                                                                                         |   |   |   |
| YEARS AT COMPANY           | Numerical Value - TOTAL NUMBER OF YEARS AT THE COMPNAY                                                                                                         |   |   |   |
| YEARS IN CURRENT ROLE      | Numerical Value -YEARS IN CURRENT ROLE                                                                                                                         |   |   |   |
| YEARS SINCE LAST PROMOTION | Numerical Value - LAST PROMOTION                                                                                                                               |   |   |   |
| YEARS WITH CURRENT MANAGER | Numerical Value - YEARS SPENT WITH CURRENT MANAGER                                                                                                             |   |   |   |

**Attition**
- Renuncia voluntaria
- Despido
- Desgaste
- Jubilación

Attrition = "Yes" → El empleado ya no trabaja en la empresa.
Attrition = "No" → El empleado aún trabaja en la empresa.

**¿Por qué es importante?**
Las empresas analizan la tasa de attrition para:
- Identificar causas de renuncias
- Prevenir la fuga de talento
- Mejorar el clima laboral
- Reducir costos por rotación

---

- Importación de módulos
- Carga de conjuntos de datos
- Procesamiento de datos
- Cálculo del tamaño del DataFrame
- Incluir las etiquetas de columna
- Generación de información básica de atributos
- Incluir características numéricas
- Incluir características categóricas
- Comprobación de valores faltantes
- Análisis descriptivo de atributos numéricos
- Eliminar columnas innecesarias
- Análisis descriptivo de atributos categóricos
- Comprobación de valores únicos en atributos categóricos
- Guardar el DataFrame en un archivo CSV

<div style="text-align: center; background-color: yellow; padding: 10px;">
    <h2 style="font-weight: bold;">IMPORTACIÓN DE MÓDULOS</h2>
</div>

In [34]:
# Libreria para la manipulación de datos
import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None)

# Libreria para modelado estadístico
from sklearn.preprocessing import LabelEncoder

# Ignorar warnings
import warnings
warnings.filterwarnings('always')
warnings.filterwarnings('ignore')

<div style="text-align: center; background-color: yellow; padding: 10px;">
    <h2 style="font-weight: bold;">CARGANDO CONJUNTO DE DATOS</h2>
</div>

In [35]:
employee_data = pd.read_csv(r'IBM-HR-Analytics-Employee-Attrition-and-Performance.csv')

In [36]:
# Imprimir las primeras 5 filas del dataframe.
employee_data.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   EnvironmentSatisfaction  Gender  HourlyRate  JobInvolvement  JobLevel  \
0                        2  Female          94               3         2   
1                        3    Male          61               2         2   
2                        4    Male          92               2         1   
3                        4  Female          56               3         1   
4                        1    Male          40               3         1   

                 JobRole  JobSatisfaction MaritalStatus  MonthlyIncome  \
0        Sales Executive                4        Single           5993   
1     Research Scientist                2       Married           5130   
2  Laboratory Technician                3        Single           2090   
3     Research Scientist                3       Married           2909   
4  Laboratory Technician                2       Married           3468   

   MonthlyRate  NumCompaniesWorked Over18 OverTime  PercentSalaryHike  \
0        19479                   8      Y      Yes                 11   
1        24907                   1      Y       No                 23   
2         2396                   6      Y      Yes                 15   
3        23159                   1      Y      Yes                 11   
4        16632                   9      Y       No                 12   

   PerformanceRating  RelationshipSatisfaction  StandardHours  \
0                  3                         1             80   
1                  4                         4             80   
2                  3                         2             80   
3                  3                         3             80   
4                  3                         4             80   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                 0                  8                      0   
1                 1                 10                      3   
2                 0                  7                      3   
3                 0                  8                      3   
4                 1                  6                      3   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                1               6                   4   
1                3              10                   7   
2                3               0                   0   
3                3               8                   7   
4                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2

In [37]:
# Imprimir las últimas 5 filas del dataframe.
employee_data.tail()

Age Attrition     BusinessTravel  DailyRate              Department  \
1465   36        No  Travel_Frequently        884  Research & Development   
1466   39        No      Travel_Rarely        613  Research & Development   
1467   27        No      Travel_Rarely        155  Research & Development   
1468   49        No  Travel_Frequently       1023                   Sales   
1469   34        No      Travel_Rarely        628  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  \
1465                23          2        Medical              1   
1466                 6          1        Medical              1   
1467                 4          3  Life Sciences              1   
1468                 2          3        Medical              1   
1469                 8          3        Medical              1   

      EmployeeNumber  EnvironmentSatisfaction Gender  HourlyRate  \
1465            2061                        3   Male          41   
1466            2062                        4   Male          42   
1467            2064                        2   Male          87   
1468            2065                        4   Male          63   
1469            2068                        2   Male          82   

      JobInvolvement  JobLevel                    JobRole  JobSatisfaction  \
1465               4         2      Laboratory Technician                4   
1466               2         3  Healthcare Representative                1   
1467               4         2     Manufacturing Director                2   
1468               2         2            Sales Executive                2   
1469               4         2      Laboratory Technician                3   

     MaritalStatus  MonthlyIncome  MonthlyRate  NumCompaniesWorked Over18  \
1465       Married           2571        12290                   4      Y   
1466       Married           9991        21457                   4      Y   
1467       Married           6142         5174                   1      Y   
1468       Married           5390        13243                   2      Y   
1469       Married           4404        10228                   2      Y   

     OverTime  PercentSalaryHike  PerformanceRating  RelationshipSatisfaction  \
1465       No                 17                  3                         3   
1466       No                 15                  3                         1   
1467      Yes                 20                  4                         2   
1468       No                 14                  3                         4   
1469       No                 12                  3                         1   

      StandardHours  StockOptionLevel  TotalWorkingYears  \
1465             80                 1                 17   
1466             80                 1                  9   
1467             80                 1                  6   
1468             80                 0                 17   
1469             80                 0                  6   

      TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
1465                      3                3               5   
1466                      5                3               7   
1467                      0                3               6   
1468                      3                2               9   
1469                      3                4               4   

      YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
1465                   2                        0                     3  
1466                   7                        1                     7  
1467                   2                        0                     3  
1468                   6                        0                     8  
1469                   3                        1                     2

<div style="text-align: center; background-color: yellow; padding: 10px;">
    <h2 style="font-weight: bold;">MANEJO DE DATOS</h2>
</div>

## <span style='color:blue'> 1] CÁLCULO DEL TAMAÑO DEL CONJUNTO DE DATOS </span>

In [38]:
# Print the shape of the DataFrame
print("Tamaño del dataframe:", employee_data.shape)
# Print the length (number of rows) of the DataFrame
print("Numero de filas del dataframe:", len(employee_data))
# Print the number of columns in the DataFrame
print("Numero de columnas del dataframe:", len(employee_data.columns))

Tamaño del dataframe: (1470, 35)
Numero de filas del dataframe: 1470
Numero de columnas del dataframe: 35


## <span style='color:blue'> 2] LISTAR COLUMNAS DEL DATAFRAME </span>

In [39]:
for column in employee_data.columns:
    print(column)

Age
Attrition
BusinessTravel
DailyRate
Department
DistanceFromHome
Education
EducationField
EmployeeCount
EmployeeNumber
EnvironmentSatisfaction
Gender
HourlyRate
JobInvolvement
JobLevel
JobRole
JobSatisfaction
MaritalStatus
MonthlyIncome
MonthlyRate
NumCompaniesWorked
Over18
OverTime
PercentSalaryHike
PerformanceRating
RelationshipSatisfaction
StandardHours
StockOptionLevel
TotalWorkingYears
TrainingTimesLastYear
WorkLifeBalance
YearsAtCompany
YearsInCurrentRole
YearsSinceLastPromotion
YearsWithCurrManager


## <span style='color:blue'> 3] GENERACIÓN DE INFORMACIÓN BÁSICA DE ATRIBUTOS </span>

In [40]:
# Imprima el resumen del dataframe estableciendo verbose = True
# Verifique valores no nulos o no nulos en el conjunto de datos.
print(employee_data.info(verbose = True))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

### <font color=red>Conclusión:</font>

1. Solo hay 26 atributos numéricos en el conjunto de datos.
2. Por otro lado, tenemos 9 atributos categóricos.

## <span style='color:blue'> 4] LISTAR CARACTERÍSTICAS NUMÉRICAS </span>

In [41]:
employee_data.select_dtypes(np.number).sample(5)

Age  DailyRate  DistanceFromHome  Education  EmployeeCount  \
825    29        718                 8          1              1   
1015   34        560                 1          4              1   
1142   45       1015                 5          5              1   
571    26        575                 1          2              1   
644    31       1222                11          4              1   

      EmployeeNumber  EnvironmentSatisfaction  HourlyRate  JobInvolvement  \
825             1150                        2          79               2   
1015            1431                        4          91               3   
1142            1611                        3          50               1   
571              792                        1          71               1   
644              895                        4          48               3   

      JobLevel  JobSatisfaction  MonthlyIncome  MonthlyRate  \
825          2                4           5056        17689   
1015         1                1           2996        20284   
1142         2                1           5769        23447   
571          1                4           4364         5288   
644          1                4           2356        14871   

      NumCompaniesWorked  PercentSalaryHike  PerformanceRating  \
825                    1                 15                  3   
1015                   5                 14                  3   
1142                   1                 14                  3   
571                    3                 14                  3   
644                    3                 19                  3   

      RelationshipSatisfaction  StandardHours  StockOptionLevel  \
825                          3             80                 1   
1015                         3             80                 2   
1142                         1             80                 0   
571                          1             80                 1   
644                          2             80                 1   

      TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  \
825                  10                      2                2   
1015                 10                      2                3   
1142                 10                      3                3   
571                   5                      2                3   
644                   8                      2                3   

      YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
825               10                   7                        1   
1015               4                   3                        1   
1142              10                   7                        1   
571                2                   2                        2   
644                6                   4                        0   

      YearsWithCurrManager  
825                      2  
1015                     3  
1142                     4  
571                      0  
644                      2

### <font color=red>Conclusión:</font>

1. Algunas de las características numéricas almacenan categorías etiquetadas con números.
2. Para un mejor análisis, reemplazaremos esos valores numéricos etiquetados con valores categóricos apropiados.

### 4.1] Etiquetado de categorías en características numéricas

In [42]:
employee_data["Education"] = employee_data["Education"].replace({1:"Below College",2:"College",3:"Bachelor",4:"Master",5:"Doctor"})

In [43]:
employee_data["EnvironmentSatisfaction"] = employee_data["EnvironmentSatisfaction"].replace({1:"Low",2:"Medium",3:"High",4:"Very High"})

In [44]:
employee_data["JobInvolvement"] = employee_data["JobInvolvement"].replace({1:"Low",2:"Medium",3:"High",4:"Very High"})

In [45]:
employee_data["JobLevel"] = employee_data["JobLevel"].replace({1:"Entry Level",2:"Junior Level",3:"Mid Level",4:"Senior Level",
                                         5:"Executive Level"})

In [46]:
employee_data["JobSatisfaction"] = employee_data["JobSatisfaction"].replace({1:"Low",2:"Medium",3:"High",4:"Very High"})

In [47]:
employee_data["PerformanceRating"] = employee_data["PerformanceRating"].replace({1:"Low",2:"Good",3:"Excellent",4:"Outstanding"})

In [48]:
employee_data["RelationshipSatisfaction"] = employee_data["RelationshipSatisfaction"].replace({1:"Low",2:"Medium",3:"High",4:"Very High"})

In [49]:
employee_data["WorkLifeBalance"] = employee_data["WorkLifeBalance"].replace({1:"Bad",2:"Good",3:"Better",4:"Best"})

## <span style='color:blue'> 5] LISTAR CARACTERÍSTICAS CATEGÓRICAS </span>

In [50]:
employee_data.select_dtypes(include="O").sample(5)

Attrition     BusinessTravel              Department Education  \
1368        No  Travel_Frequently  Research & Development    Master   
1100        No      Travel_Rarely                   Sales    Master   
84          No      Travel_Rarely  Research & Development   College   
156         No      Travel_Rarely  Research & Development    Master   
596         No      Travel_Rarely  Research & Development    Master   

     EducationField EnvironmentSatisfaction  Gender JobInvolvement  \
1368          Other                    High    Male         Medium   
1100  Life Sciences                  Medium  Female         Medium   
84          Medical                     Low    Male           High   
156         Medical                  Medium    Male         Medium   
596   Life Sciences               Very High  Female      Very High   

          JobLevel                 JobRole JobSatisfaction MaritalStatus  \
1368  Junior Level      Research Scientist       Very High       Married   
1100   Entry Level    Sales Representative            High       Married   
84    Junior Level  Manufacturing Director          Medium       Married   
156   Junior Level  Manufacturing Director            High       Married   
596    Entry Level      Research Scientist            High        Single   

     Over18 OverTime PerformanceRating RelationshipSatisfaction  \
1368      Y      Yes         Excellent                   Medium   
1100      Y       No       Outstanding                      Low   
84        Y       No         Excellent                     High   
156       Y       No         Excellent                     High   
596       Y       No         Excellent                     High   

     WorkLifeBalance  
1368          Better  
1100          Better  
84            Better  
156           Better  
596           Better

## <span style='color:blue'> 6] COMPROBAR VALORES FALTANTES </span>

In [51]:
# Calcular el número de valores faltantes en cada columna
    
missing_df = employee_data.isnull().sum().to_frame().rename(columns={0:"Total No. of Missing Values"})
missing_df["% of Missing Values"] = round((missing_df["Total No. of Missing Values"]/len(employee_data))*100,2)
missing_df

Total No. of Missing Values  % of Missing Values
Age                                                 0                  0.0
Attrition                                           0                  0.0
BusinessTravel                                      0                  0.0
DailyRate                                           0                  0.0
Department                                          0                  0.0
DistanceFromHome                                    0                  0.0
Education                                           0                  0.0
EducationField                                      0                  0.0
EmployeeCount                                       0                  0.0
EmployeeNumber                                      0                  0.0
EnvironmentSatisfaction                             0                  0.0
Gender                                              0                  0.0
HourlyRate                                          0                  0.0
JobInvolvement                                      0                  0.0
JobLevel                                            0                  0.0
JobRole                                             0                  0.0
JobSatisfaction                                     0                  0.0
MaritalStatus                                       0                  0.0
MonthlyIncome                                       0                  0.0
MonthlyRate                                         0                  0.0
NumCompaniesWorked                                  0                  0.0
Over18                                              0                  0.0
OverTime                                            0                  0.0
PercentSalaryHike                                   0                  0.0
PerformanceRating                                   0                  0.0
RelationshipSatisfaction                            0                  0.0
StandardHours                                       0                  0.0
StockOptionLevel                                    0                  0.0
TotalWorkingYears                                   0                  0.0
TrainingTimesLastYear                               0                  0.0
WorkLifeBalance                                     0                  0.0
YearsAtCompany                                      0                  0.0
YearsInCurrentRole                                  0                  0.0
YearsSinceLastPromotion                             0                  0.0
YearsWithCurrManager                                0                  0.0

### <font color=red>Conclusión:</font>

1. Ninguno de los atributos presenta valores faltantes.
2. Dado que no hay valores faltantes, nuestro análisis posterior será consistente e imparcial.

## <span style='color:blue'> 7] ANÁLISIS DESCRIPTIVO DE ATRIBUTOS NUMÉRICOS </span>

In [21]:
employee_data.describe().T

count          mean          std     min      25%  \
Age                      1470.0     36.923810     9.135373    18.0    30.00   
DailyRate                1470.0    802.485714   403.509100   102.0   465.00   
DistanceFromHome         1470.0      9.192517     8.106864     1.0     2.00   
EmployeeCount            1470.0      1.000000     0.000000     1.0     1.00   
EmployeeNumber           1470.0   1024.865306   602.024335     1.0   491.25   
HourlyRate               1470.0     65.891156    20.329428    30.0    48.00   
MonthlyIncome            1470.0   6502.931293  4707.956783  1009.0  2911.00   
MonthlyRate              1470.0  14313.103401  7117.786044  2094.0  8047.00   
NumCompaniesWorked       1470.0      2.693197     2.498009     0.0     1.00   
PercentSalaryHike        1470.0     15.209524     3.659938    11.0    12.00   
StandardHours            1470.0     80.000000     0.000000    80.0    80.00   
StockOptionLevel         1470.0      0.793878     0.852077     0.0     0.00   
TotalWorkingYears        1470.0     11.279592     7.780782     0.0     6.00   
TrainingTimesLastYear    1470.0      2.799320     1.289271     0.0     2.00   
YearsAtCompany           1470.0      7.008163     6.126525     0.0     3.00   
YearsInCurrentRole       1470.0      4.229252     3.623137     0.0     2.00   
YearsSinceLastPromotion  1470.0      2.187755     3.222430     0.0     0.00   
YearsWithCurrManager     1470.0      4.123129     3.568136     0.0     2.00   

                             50%       75%      max  
Age                         36.0     43.00     60.0  
DailyRate                  802.0   1157.00   1499.0  
DistanceFromHome             7.0     14.00     29.0  
EmployeeCount                1.0      1.00      1.0  
EmployeeNumber            1020.5   1555.75   2068.0  
HourlyRate                  66.0     83.75    100.0  
MonthlyIncome             4919.0   8379.00  19999.0  
MonthlyRate              14235.5  20461.50  26999.0  
NumCompaniesWorked           2.0      4.00      9.0  
PercentSalaryHike           14.0     18.00     25.0  
StandardHours               80.0     80.00     80.0  
StockOptionLevel             1.0      1.00      3.0  
TotalWorkingYears           10.0     15.00     40.0  
TrainingTimesLastYear        3.0      3.00      6.0  
YearsAtCompany               5.0      9.00     40.0  
YearsInCurrentRole           3.0      7.00     18.0  
YearsSinceLastPromotion      1.0      3.00     15.0  
YearsWithCurrManager         3.0      7.00     17.0

### <font color=red>CONCLUSION:</font>

1. La edad mínima es 18 años, lo que indica que todos los empleados son adultos. Por lo tanto, no es necesario el atributo "Mayores de 18" para nuestro análisis.
2. El valor de la desviación estándar de "EmployeeCount" y "StandardHours" es 0,00, lo que indica que todos los valores presentes en este atributo son iguales.
3. El atributo "EmployeeNumber" representa un valor único para cada empleado, lo que no proporciona información significativa.
4. Dado que este atributo no proporciona información significativa para nuestro análisis, podemos simplemente omitirlo.

## <span style='color:blue'> 8] ELIMINA LAS COLUMNAS INNECESARIAS </span>

**Informe de observación:** _Observamos que "EmployeeCount", "Over18" y "StandardHours" tienen solo un valor único, mientras que "EmployeeNumber" tiene 1470 valores únicos. Estas funciones no nos resultan útiles, por lo que eliminaremos esas columnas._

In [52]:
employee_data.drop(['EmployeeCount', 'EmployeeNumber', 'Over18', 'StandardHours'], axis="columns", inplace=True)

In [54]:
# Mostrar 5 primeras filas del dataframe
employee_data.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome      Education EducationField EnvironmentSatisfaction  \
0                 1        College  Life Sciences                  Medium   
1                 8  Below College  Life Sciences                    High   
2                 2        College          Other               Very High   
3                 3         Master  Life Sciences               Very High   
4                 2  Below College        Medical                     Low   

   Gender  HourlyRate JobInvolvement      JobLevel                JobRole  \
0  Female          94           High  Junior Level        Sales Executive   
1    Male          61         Medium  Junior Level     Research Scientist   
2    Male          92         Medium   Entry Level  Laboratory Technician   
3  Female          56           High   Entry Level     Research Scientist   
4    Male          40           High   Entry Level  Laboratory Technician   

  JobSatisfaction MaritalStatus  MonthlyIncome  MonthlyRate  \
0       Very High        Single           5993        19479   
1          Medium       Married           5130        24907   
2            High        Single           2090         2396   
3            High       Married           2909        23159   
4          Medium       Married           3468        16632   

   NumCompaniesWorked OverTime  PercentSalaryHike PerformanceRating  \
0                   8      Yes                 11         Excellent   
1                   1       No                 23       Outstanding   
2                   6      Yes                 15         Excellent   
3                   1      Yes                 11         Excellent   
4                   9       No                 12         Excellent   

  RelationshipSatisfaction  StockOptionLevel  TotalWorkingYears  \
0                      Low                 0                  8   
1                Very High                 1                 10   
2                   Medium                 0                  7   
3                     High                 0                  8   
4                Very High                 1                  6   

   TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                      0             Bad               6                   4   
1                      3          Better              10                   7   
2                      3          Better               0                   0   
3                      3          Better               8                   7   
4                      3          Better               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2

In [55]:
print("tamaño del dataFrame:", employee_data.shape)
print("Numero de filas del dataframe:", len(employee_data))
print("Numero de filas del dataframe:", len(employee_data.columns))

tamaño del dataFrame: (1470, 31)
Numero de filas del dataframe: 1470
Numero de filas del dataframe: 31


In [56]:
for column in employee_data.columns:
    print(column)

Age
Attrition
BusinessTravel
DailyRate
Department
DistanceFromHome
Education
EducationField
EnvironmentSatisfaction
Gender
HourlyRate
JobInvolvement
JobLevel
JobRole
JobSatisfaction
MaritalStatus
MonthlyIncome
MonthlyRate
NumCompaniesWorked
OverTime
PercentSalaryHike
PerformanceRating
RelationshipSatisfaction
StockOptionLevel
TotalWorkingYears
TrainingTimesLastYear
WorkLifeBalance
YearsAtCompany
YearsInCurrentRole
YearsSinceLastPromotion
YearsWithCurrManager


## <span style='color:blue'> 9] ANÁLISIS DESCRIPTIVO DE ATRIBUTOS CATEGÓRICOS </span>

In [57]:
employee_data.describe(include="O").T

count unique                     top  freq
Attrition                 1470      2                      No  1233
BusinessTravel            1470      3           Travel_Rarely  1043
Department                1470      3  Research & Development   961
Education                 1470      5                Bachelor   572
EducationField            1470      6           Life Sciences   606
EnvironmentSatisfaction   1470      4                    High   453
Gender                    1470      2                    Male   882
JobInvolvement            1470      4                    High   868
JobLevel                  1470      5             Entry Level   543
JobRole                   1470      9         Sales Executive   326
JobSatisfaction           1470      4               Very High   459
MaritalStatus             1470      3                 Married   673
OverTime                  1470      2                      No  1054
PerformanceRating         1470      2               Excellent  1244
RelationshipSatisfaction  1470      4                    High   459
WorkLifeBalance           1470      4                  Better   893

### <font color=red>CONCLUSION:</font>

Todos los atributos categóricos tienen baja cardinalidad.
(Es decir, tienen pocos valores distintos posibles.)

Las columnas **Attrition** y **OverTime** están altamente sesgadas hacia la categoría 'No'.
(La mayoría de los registros tienen el valor 'No' en esas columnas.)

El atributo **BusinessTravel** está altamente sesgado hacia la categoría **Travel_Rarely**.
(La mayoría de los empleados viajan rara vez por trabajo.)

## <span style='color:blue'> 10] COMPROBACIÓN DEL VALOR ÚNICO DE LOS ATRIBUTOS CATEGÓRICOS </span>

In [58]:
# Calcular el número de valores únicos en cada columna
for column in employee_data.columns:
    print(f"{column} - Número de valores únicos : {employee_data[column].nunique()}")
    print("=============================================================")

Age - Número de valores únicos : 43
Attrition - Número de valores únicos : 2
BusinessTravel - Número de valores únicos : 3
DailyRate - Número de valores únicos : 886
Department - Número de valores únicos : 3
DistanceFromHome - Número de valores únicos : 29
Education - Número de valores únicos : 5
EducationField - Número de valores únicos : 6
EnvironmentSatisfaction - Número de valores únicos : 4
Gender - Número de valores únicos : 2
HourlyRate - Número de valores únicos : 71
JobInvolvement - Número de valores únicos : 4
JobLevel - Número de valores únicos : 5
JobRole - Número de valores únicos : 9
JobSatisfaction - Número de valores únicos : 4
MaritalStatus - Número de valores únicos : 3
MonthlyIncome - Número de valores únicos : 1349
MonthlyRate - Número de valores únicos : 1427
NumCompaniesWorked - Número de valores únicos : 10
OverTime - Número de valores únicos : 2
PercentSalaryHike - Número de valores únicos : 15
PerformanceRating - Número de valores únicos : 2
RelationshipSatisfa

In [59]:
categorical_features = []
for column in employee_data.columns:
    if employee_data[column].dtype == object and len(employee_data[column].unique()) <= 30:
        categorical_features.append(column)
        print(f"{column} : {employee_data[column].unique()}")
        print(employee_data[column].value_counts())
        print("====================================================================================")
categorical_features.remove('Attrition')

Attrition : ['Yes' 'No']
Attrition
No     1233
Yes     237
Name: count, dtype: int64
BusinessTravel : ['Travel_Rarely' 'Travel_Frequently' 'Non-Travel']
BusinessTravel
Travel_Rarely        1043
Travel_Frequently     277
Non-Travel            150
Name: count, dtype: int64
Department : ['Sales' 'Research & Development' 'Human Resources']
Department
Research & Development    961
Sales                     446
Human Resources            63
Name: count, dtype: int64
Education : ['College' 'Below College' 'Master' 'Bachelor' 'Doctor']
Education
Bachelor         572
Master           398
College          282
Below College    170
Doctor            48
Name: count, dtype: int64
EducationField : ['Life Sciences' 'Other' 'Medical' 'Marketing' 'Technical Degree'
 'Human Resources']
EducationField
Life Sciences       606
Medical             464
Marketing           159
Technical Degree    132
Other                82
Human Resources      27
Name: count, dtype: int64
EnvironmentSatisfaction : ['Medium' '

### <font color=red>CONCLUSION:</font>

1. El conjunto de valores de los atributos categóricos es completo y fácil de comprender.
2. Por lo tanto, no es necesario realizar pasos de preprocesamiento para estos atributos.

<div style="text-align: center; background-color: yellow; padding: 10px;">
    <h2 style="font-weight: bold;">GUARDAR EL DATAFRAME EN UN ARCHIVO CSV</h2>
</div>

In [62]:
# Guardar como CSV
employee_data.to_csv('IBM-HR-Analytics-Employee-Attrition-and-Performance-Revised.csv', index=False)

# Guardar como Excel
employee_data.to_excel('IBM-HR-Analytics-Employee-Attrition-and-Performance-Revised.xlsx', index=False)